### DISCUSSION TITLE
Partial object schema should allow for partial success

### DISCUSSION THREAD

**Discussion by JacobWeisenburger** [colinhacks/zod#875] [Q&A]
I feel like if I have an object schema that I mark as partial, then I should be able to get the part of the data that was successfully parsed.
```ts
const schema = z.object( {
    foo: z.literal( 'foo' ),
    bar: z.string(),
} ).partial()
```

However, when I parse the data, all of it needs to pass the parse or else it's an error.

```ts
schema.safeParse( { foo: 'foo', bar: 'blah blah' } ) // success
schema.safeParse( { foo: 'foo' } ) // success
schema.safeParse( { foo: 'foo', bar: 2 } ) // error
```

This seems not great to me because if the user is filling out a form and I'm using the schema for validation, then when they submit the form, if anything is wrong, then all the data is erased instead of just the fields that were wrong.

___

I would like something like this:
```ts
schema.safeParse( { foo: 'foo', bar: 2 } )
// {
//     partialSuccess: true,
//     validData: { foo: 'foo' },
//     invalidData: { bar: 2 },
//     error: { ... }
// }
```

Is there some way to do this?

**Comment by StefanTerdell** [comment]
I don't think so. Since the output data can be transformed, it is not safe to return input data as partial success data, and if Zod hits an error there is currently no way to retrieve the parts of the data that was successful in its validated and potentially transformed state. 

However, in the case of this form validation, that shouldn't really be necessary. I think your problems are probably due to the form validation implementation. Yeeting the entire form state due to a validation error seems odd to me, TBH. You can find which fields have errors through `safeParseResult.error.errors.map(({path}) => path[0])` after all, and - depending on the implementation - clear or mark only those fields. I presume you're not really interested in the transformed data due to the errors still blocking the submit, so it shouldn't matter.

Or am I missing some context? 🤔

**Reply by JacobWeisenburger** [reply]
https://github.com/colinhacks/zod/discussions/945

See, I'm not the only one that wants partial success. :D

**Reply by StefanTerdell** [reply]
Fair enough, but I'm still not convinced it's a great idea. How would such an output be typed? Always as partial? The alternative is that you'd get unreliable type declarations.

Come to think of it, perhaps an extension such as `.fallible()` could be helpful and integrate well into the dev API. It could generate an `.optional()` type and provide an error context, returning undefined when catching bad data... With the way the error handling is done right now I'm not sure its possible to implement though, but could certainly have a place on the v4 wish list.

**Reply by JacobWeisenburger** [reply]
Good enough for me 😎😀
